### PR TITLE
Fix hang on Windows in x509 policy code

### LINKED ISSUE #19643: PKITS 4.10.7 and 4.10.8 hang on Windows in OpenSSL 3.0
Presuming `TA.pem` is the trust anchor from PKITS, the following works fine on GNU/Linux but hangs on Windows:

```
% openssl cms -verify -CAfile TA.pem -policy anyPolicy -in SignedInvalidMappingFromanyPolicyTest7.eml
CMS Verification failure
0046881801000000:error:17000064:CMS routines:cms_signerinfo_verify_cert:certificate verify error:../crypto/cms/cms_smime.c:289:Verify error: invalid or inconsistent certificate policy extension
```

Similarly for the next test.

git bisect suggests the problem is 9aa4be691f5c73eb3c68606d824c104550c053f7 so I presume the difference is something to do with the way locking works on Windows compared to pthreads.

### PR DESCRIPTION
**Pull request by paulidale**
This reverts commit 9aa4be691f5c73eb3c68606d824c104550c053f7.

Fixes #19643

### CODE REVIEW

**Comment by t8m** [comment]
A test will probably work only in tsan or on Windows. But yeah, it would be useful.

File: crypto/x509/pcy_map.c [hunk 1]
```diff
@@ -73,10 +73,8 @@ int ossl_policy_cache_set_mapping(X509 *x, POLICY_MAPPINGS *maps)
 
     ret = 1;
  bad_mapping:
-    if (ret == -1 && CRYPTO_THREAD_write_lock(x->lock)) {
+    if (ret == -1)
         x->ex_flags |= EXFLAG_INVALID_POLICY;
```
Comment (mattcaswell) [inline crypto/x509/pcy_map.c:77]: You could even remove this completely since `policy_cache_new` does it again anyway.

Comment (t8m) [reply crypto/x509/pcy_map.c:77]: yeah, that would remove the coverity as well.

**Comment by paulidale** [comment]
Testcase added.

I suspect getting the `test/pkits-test.pl` script running as a unit test would be better.  I've raised this as #19663.

**Review by t8m — DISMISSED** [review]
Please amend the first commit message because the commit is no longer just revert but it completely drops the unnecessary ex_flags setting.

**Comment by mattcaswell** [comment]
Are there licence and/or CLA issues with just including the PKITS test data in our codebase?

**Comment by paulidale** [comment]
They come from NIST which generally mean fair game to use.  We've a lot of other NIST test cases already.

**Comment by mattcaswell** [comment]
> They come from NIST which generally mean fair game to use.

We should check this with OMC.

File: test/recipes/90-test_threads.t [hunk 2]
```diff
@@ -54,3 +54,14 @@ close CFGOUT;
...
+# Test case for the locking problem reported in #19643.
+# This will fail if the fix is in and deadlock on Windows (and possibly
+# other platforms) if not.
+ok(!run(app(['openssl', 'cms', '-verify',
+             '-CAfile', srctop_file("test/certs", "pkitsta.pem"),
+             '-policy', 'anyPolicy',
+             '-in', srctop_file("test/smime-eml",
+                                "SignedInvalidMappingFromanyPolicyTest7.eml")
+            ])),
+   "issue#19643");
```
Comment (t8m) [inline test/recipes/90-test_threads.t:67]: I would put this into test_cms. I suppose the test won't work with no-cms.

**Comment by mattcaswell** [comment]
OMC: Inclusion of this test data is ok to proceed.

**Comment by t-j-h** [comment]
FYI https://csrc.nist.gov/projects/pki-testing is the web page from which the test data is located.

**Comment by paulidale** [comment]
Relocated.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
ping for second review

**Review by beldmit — APPROVED** [review]
LGTM

File: crypto/x509/pcy_map.c [hunk 3]
```diff
@@ -73,10 +73,6 @@ int ossl_policy_cache_set_mapping(X509 *x, POLICY_MAPPINGS *maps)
 
     ret = 1;
  bad_mapping:
-    if (ret == -1 && CRYPTO_THREAD_write_lock(x->lock)) {
-        x->ex_flags |= EXFLAG_INVALID_POLICY;
```
Comment (FdaSilvaYY) [inline crypto/x509/pcy_map.c:77]: This line of code was presented in original code version, no ?
The version before the reverted commit ?

**Comment by t8m** [comment]
Merged to master, 3.1, and 3.0 branches. Thank you.

**Pull request closed without merging**